syntax = "proto3";

service User {
  rpc LocalStore(LocalStoreMessage) returns (Result) {}
  rpc CreateChain(CreateChainMessage) returns (Result) {}
  rpc ListChain(ListChainMessage) returns (ChainResult) {}
  rpc WriteOperation(WriteOperationMessage) returns (WriteOperationResult) {}
  rpc ListBooks(ListBooksMessage) returns (ListBooksResult) {}
  rpc ReadOperation(ReadOperationMessage) returns (ReadOperationResult) {}
  rpc TimeOut(TimeOutMessage) returns (Result) {}
  rpc DataStatus(DataStatusMessage) returns (DataStatusResult) {}
  rpc RemoveHead(RemoveHeadMessage) returns (ChainResult) {}
  rpc RestoreHead(RestoreHeadMessage) returns (ChainResult) {}
  rpc Ping(Empty) returns (Empty) {}
  rpc GetProcesses(Empty) returns (ProcessList) {}
  rpc CreateProcesses(CreateProcessesMessage) returns (Empty) {}
  rpc UpdateProcesses(UpdateMessage) returns (Empty) {}

}

message UpdateMessage {
  string previous = 1;
  string current = 2;
  string next = 3;
  string head = 4;
  string tail = 5;
}

message ProcessList {
  repeated string processes = 1;
}

message Empty {}

message LocalStoreMessage {
  string temp = 1;
}

message CreateChainMessage {
  string temp = 1;
}

message ListChainMessage {
  string path = 1;
  string next = 2;
}

message WriteOperationMessage {
  string temp = 1;
}

message ListBooksMessage {
  string temp = 1;
}

message ReadOperationMessage {
  string temp = 1;
}

message TimeOutMessage {
  string temp = 1;
}

message DataStatusMessage {
  string temp = 1;
}

message RemoveHeadMessage {
  string temp = 1;
}

message RestoreHeadMessage {
  string temp = 1;
}


message Result {
  bool success = 1;
}

message ChainResult {
  string chain = 1;
}

message WriteOperationResult {
  string bookNameAndPrice = 1;
}

message ListBooksResult {
  string booksList = 1;
}

message ReadOperationResult {
  string bookPrice = 1;
}

message DataStatusResult {
  string booksStatus = 1;
}

message CreateProcessesMessage {
  int32 amount = 1;
}